### PR TITLE
Contracts & Harnesses for `NonNull::read`, `NonNull::read_volatile`, `NonNull::read_unaligned`

### PR DESCRIPTION
**Pull request by QinyuanWu**
Towards #52 

Contracts & Harnesses for `NonNull::read`, `NonNull::read_volatile`, `NonNull::read_unaligned`
### Discussion
For the ensures clause in the contract for `NonNull::read`, I could not directly compare the value pointed by self and result due to unsized and lack of PartialEq implementation, and the comparison is currently moved to the harness.
### Verification
Successful verification:
```
SUMMARY:
 ** 0 of 141 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.1760525s

Complete - 5 successfully verified harnesses, 0 failures, 5 total.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:909]: `can_dereference` already checks that the pointer is aligned.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:910]: You can remove the `ensures` clause. Since there is no `modifies` clause in this contract, Kani will check that the value is not modified.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:932]: Remove `is_aligned`.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:933]: Drop the `ensures`.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:954]: Use `can_read_unaligned`

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1853]: This assumption is not needed.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1851]: Use upper case for constants.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1856]: The `PointerGenerator` should be more suitable for usage here.

**Comment by QinyuanWu** [comment]
Addressed previous PR comments and Increased SIZE:
```
Checking harness ptr::non_null::verify::non_null_check_read_unaligned...
SUMMARY:
 ** 0 of 261 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 18.91408s

Checking harness ptr::non_null::verify::non_null_check_read_volatile...
SUMMARY:
 ** 0 of 237 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 15.834378s
Checking harness ptr::non_null::verify::non_null_check_read...
SUMMARY:
 ** 0 of 229 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 38.454617s

Complete - 3 successfully verified harnesses, 0 failures, 3 total.
```

**Comment by QinyuanWu** [comment]
@zhassan-aws Thank you! Waiting for another review.

**Review by zhassan-aws — APPROVED** [review]
Can you please resolve the conflict?

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1898]: FYI, the PointerGenerator will generate unaligned pointers.

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1905]: I don't think we need this

**Review by celinval — APPROVED** [review]

**Comment by Eh2406** [comment]
I've then asked to help review this.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1905]: @celinval should we check whether `read_unaligned` reads the value correctly?

**Comment by HuStmpHrrr** [comment]
LGTM

**Review by HuStmpHrrr — APPROVED** [review]

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1898]: Thanks for the suggestion! Changed to PointerGenerator.

Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1898]: Thanks for the suggestion! Changed to `PointerGenerator`.

**Comment by QinyuanWu** [comment]
@zhassan-aws resolved conflicts and addressed all comments, ready to merge:
```
VERIFICATION:- SUCCESSFUL
Verification Time: 0.173812s

Complete - 9 successfully verified harnesses, 0 failures, 9 total.
```

**Comment by zhassan-aws** [comment]
trigger approval workflow

**Comment by zhassan-aws** [comment]
trigger approval workflow

**Pull request merged**
